src/community/geofence-server/src/main/java/org/geoserver/geofence/web/GeofenceRulePage.java

```diff
@@ -0,0 +1,259 @@
+                ((ShortRule) form.getModelObject()).setRequest(null);
+                requestChoice.modelChanged();
+                target.addComponent(requestChoice);
+            }
+        });
+        serviceChoice.setNullValid(true);
+        
+        form.add(requestChoice = new DropDownChoice<String>("request", getOperationNames(rule.getService()),
+                new CaseConversionRenderer()));
+        requestChoice.setOutputMarkupId(true);
+        requestChoice.setNullValid(true);
+        
+        form.add(workspaceChoice = new DropDownChoice<String>("workspace", getWorkspaceNames()));
+        workspaceChoice.add(new OnChangeAjaxBehavior() {
+            @Override
+            protected void onUpdate(AjaxRequestTarget target) {
+                layerChoice.setChoices(getLayerNames(workspaceChoice.getConvertedInput()));
+                ((ShortRule) form.getModelObject()).setLayer(null);
+                layerChoice.modelChanged();
+                target.addComponent(layerChoice);
+            }
+        });
+        workspaceChoice.setNullValid(true);
+
+        form.add(layerChoice = new DropDownChoice<String>("layer", getLayerNames(rule.getWorkspace())));
+        layerChoice.setOutputMarkupId(true);
+        layerChoice.setNullValid(true);
+
+        form.add(grantTypeChoice = 
+            new DropDownChoice<GrantType>("access", Arrays.asList(GrantType.values()), new GrantTypeRenderer()));        
+        grantTypeChoice.setRequired(true);
+
+
+        // build the submit/cancel
+        form.add(new SubmitLink("save") {
+            @Override
+            public void onSubmit() {
+                ShortRule rule = (ShortRule) getForm().getModelObject();
+                try {
+                    rules.save(rule);
+                    doReturn(GeofenceServerPage.class);
+                } catch (DuplicateKeyException e) {
+                    error(new ResourceModel("GeofenceRulePage.duplicate").getObject());
+                } catch (Exception e){
+                    error(e);
+                }
+            }
+        });
+        form.add(new BookmarkablePageLink<ShortRule>("cancel", GeofenceServerPage.class));
+    }
+
+
+    /**
+     * Returns a sorted list of workspace names
+     */
+    protected List<String> getWorkspaceNames() {
+        SortedSet<String> resultSet = new TreeSet<String>();
+        for (WorkspaceInfo ws : getCatalog().getWorkspaces()) {
+            resultSet.add(ws.getName());
+        }
+        return new ArrayList(resultSet);
+    }
+
+    /**
+     * Returns a sorted list of layer names in the specified workspace (or * if the workspace is *)
+     */
+    protected List<String> getLayerNames(String workspaceName) {
+        SortedSet<String> resultSet = new TreeSet<String>();
+        if (workspaceName != null) {
+            for (ResourceInfo r : getCatalog().getResources(ResourceInfo.class)) {
```

Comment:
Thanks Kevin, I think we need a page in our developers guide (or a round of deprecations in the javadocs) to prevent extensions casually wrecking scalability.